[GAP] - Role Management UI

legrego opened the pull request:
![image](https://user-images.githubusercontent.com/3493255/51063094-72310080-15c7-11e9-9274-debf0e9b84f1.png)

![image](https://user-images.githubusercontent.com/3493255/51063108-8aa11b00-15c7-11e9-92fc-31c922086e05.png)


![image](https://user-images.githubusercontent.com/3493255/51063165-cb992f80-15c7-11e9-894d-630c109211d3.png)






---------
Edge-case scenarios:
  1) [x] '*' and spaces in the same "entry"
Handled via `_transform_errors` at the API level. Renders a partial read-only view in the UI.
  2) [x] same space appearing in multiple "entries"
Handled via `_transform_errors` at the API level. Renders a partial read-only view in the UI.
  3) [x] base and feature privileges being set on the same "entry"
UI does not allow this to be set, but UI is smart enough to display the correct effective privilege in this case.
  4) [x] multiple base privileges set in the same "entry"
UI does not allow this to be set, but UI is smart enough to apply the most permissive base privilege when displaying and performing privilege calculations.
  5) [x] multiple feature privileges for the same "entry" (ml_all and ml_read)
UI does not allow for this to be set, but UI is smart enough to apply the most permissive base privilege when displaying and performing privilege calculations.


--------

## Summary
This updates the role management UI to allow application privileges to be customized globally and per-space.

## TODO:
- [x] [First Design review](https://github.com/legrego/kibana/pull/13)
- [ ] Second Design review
- [ ] Copy review
- [x] i18n
- [x] Handle deleted/unknown features
- [x] Handle deleted/unknown spaces
- [x] Cleanup & refactoring
- [x] Testing

legrego committed:
hacking away at privileges

legrego committed:
Merge branch 'granular-app-privileges' into gap/security-ui

legrego committed:
additional ui testing

legrego committed:
Merge branch 'granular-app-privileges' into gap/security-ui

legrego committed:
less-than-mvp, but functional against updated role management api

legrego committed:
Merge branch 'granular-app-privileges' into gap/security-ui

legrego committed:
fix saved object types in oss feature privilege registry

legrego committed:
allowed space minimum should not account for global feature toggles

legrego committed:
tooltip support for feature privileges that support it

elasticmachine commented:
Pinging @elastic/kibana-security

legrego committed:
remove dead code & fix TypeScript errors

legrego committed:
distinguish between create and edit space privileges

legrego committed:
update space privilege form to handle multiple spaces

legrego committed:
Merge branch 'granular-app-privileges' into gap/security-ui

legrego committed:
Merge branch 'granular-app-privileges' into gap/security-ui

legrego commented:
Jenkins, for kicks, test this

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3214/)

legrego committed:
enable customize privilege for UI

legrego committed:
Merge branch 'granular-app-privileges' into gap/security-ui

legrego committed:
Merge branch 'granular-app-privileges' into gap/security-ui

legrego committed:
Merge branch 'granular-app-privileges' into gap/security-ui

legrego committed:
wip rewrite of effective privileges calculations

legrego committed:
Merge branch 'granular-app-privileges' into gap/security-ui

legrego committed:
some cleanup and testing

legrego committed:
enables role management feature toggles with spaces disabled

legrego committed:
additional tests and bugfixes

legrego committed:
design feedback for privilege form

legrego committed:
remove debug code

legrego committed:
Make privilege matrix a modal

legrego committed:
remove unused components as a result of the redesign

legrego committed:
restructure kibana privilege components

legrego committed:
extracting privilege space table into its own component again

legrego committed:
additional cleanup & existing unit tests passing

legrego commented:
@elastic/kibana-design ignore the review ping, this ain't ready for you yet 😄

legrego committed:
subdue feature name when no privilege is assigned

legrego committed:
more cleanup and typesafety

legrego committed:
privilege matrix improvements

legrego committed:
consolidate rendering of privilege values into a component

legrego committed:
much refactoring; such cleanup

legrego committed:
show when privileges have been superceded by other privileges

legrego committed:
more effective privilege testing

legrego committed:
testing the factory

legrego committed:
fix typo

legrego committed:
comment and color cleanup

legrego commented:
@cchaos, This is ready for a design review when you get a chance, with a couple of notes:

1) I haven't done any accessibility testing or i18n work yet. I planned on tackling these after the design review is finished, once the markup is relatively stable.
2) I'm sure some wording will need to be tweaked. I'll get this started after as well.
3) The UI only works when creating a new role. Editing existing roles isn't supported by the UI yet. You also can't save your new role.

legrego committed:
make typescript a little happier

legrego committed:
additional testing

legrego committed:
remove 'kibana.global' property from role definition

cchaos commented:
PR4U: https://github.com/legrego/kibana/pull/13

legrego committed:
update UI to conform to new API spec

legrego committed:
Prevent both feature and base privileges from being saved for a given privilege entry

cchaos committed:
Design Edits (#13)

Here are the design edits (I felt comfortable making). I also had to add some `// TODO`’s for TS and/or EUI fixes.

There are still some issues with the UI that I'll document in here.

----

## 1. I cleaned up the "No access" screen but should get a review from writers

<img src="https://d.pr/free/i/tc7Xqj+" />

## 2. The feature list needs to be updated to use button groups.

It should only display the buttons on "Custom" and each feature should always show all privileges, disabling the ones that aren't available. Only remove the button group if the feature's privilege cannot be changed at all and show a lock icon with a tooltip.

_Forget about the bolding of **All** for now__

<img width="2036" alt="screen shot 2019-01-08 at 14 39 14 pm" src="https://user-images.githubusercontent.com/549577/50920616-d90db880-1413-11e9-93cf-5575887f7024.png">

## 3. The "base" privilege selection should also allow a "None" option which behaves similar to the "Read" and "All".

<img src="https://d.pr/free/i/0fLOpU+" />

**Bug:**
<img src="https://d.pr/free/i/WR9QW9+" />

## 4. I don't understand the ML tooltip

This could use more information or better instruction.

<img src="https://d.pr/free/i/fT8xng+" />

## 5. Anywhere a permission is superceded by the Global permissions, it should show a lock with a tooltip

<img src="https://d.pr/free/i/0f8z04+" />

<img src="https://d.pr/free/i/Cv5fV8+" />

<img src="https://d.pr/free/i/lXoDvg+" />

*Question*
<img src="https://d.pr/free/i/mAjYlc+" />

## 6. Let's lock the privilege select select dropdown if super-ceded by global

<img src="https://d.pr/free/i/Q6HgYm+" />

## 7. Functionality around the`+7 more` text

**In the regular table**, lets just allow the row to expand to show the entire list.

<img src="https://d.pr/free/i/mD3Nh2+" />

**In the Matrix**, let's make the text a popover link to show the full list (including those visible in the header)

<img src="https://d.pr/free/i/bGJvag+" />

**For Global (in matrix and regular table)**, let's do the same thing as above but make the `(all spaces)` a popover link that lists all spaces available in the whole Kibana instance.

<img src="https://d.pr/free/i/mQ6icn+" />

----

😌 Sorry, that was probably long-winded, but I think should cover it.

Feel free, if you want, to commit directly to my forked branch before you merge in.

legrego committed:
Merge branch 'gap/security-ui' of github.com:legrego/kibana into gap/security-ui

legrego committed:
Merge branch 'granular-app-privileges' into gap/security-ui

legrego committed:
Change feature privilege selection to EuiButtonGroup

legrego committed:
Allow feature privileges to be changed in bulk

legrego committed:
improve tooltip messages

legrego committed:
Show lock any time a privilege is set by global privileges

legrego committed:
allow privilege table to expand/collapse spaces in each group

legrego committed:
stubbing i18n for the PrivilegeMatrix

legrego committed:
show all impacted spaces in popover inside PrivilegeMatrix

legrego committed:
Merge branch 'granular-app-privileges' into gap/security-ui

legrego committed:
support editing existing roles

legrego committed:
updates to support existing functional tests

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6798/)

legrego committed:
snapshot updates

legrego committed:
update imports

legrego committed:
making typescript happy

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6842/)

legrego committed:
Update i18n id prefixes

legrego committed:
Merge branch 'granular-app-privileges' into gap/security-ui

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6885/)

legrego committed:
i18n fixes

legrego committed:
support deleted spaces when editing roles

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6912/)

legrego committed:
adds read-only view for roles with a transform error

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6937/)

legrego committed:
handle scenario where multiple base privileges are assigned

legrego committed:
update test

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6943/)

legrego committed:
fix ts

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6950/)

legrego committed:
consolidate PrivilegeMap definition

legrego committed:
make use of FeaturePrivilegeSet type where possible

legrego committed:
cleanup and consolidate imports from 'common/model'

legrego committed:
rename role.ts => role_utils.ts

legrego committed:
move copyRole into role_utils

legrego committed:
consolidate logic for determining the global privilege definition

legrego committed:
add link to show all spaces when viewing global privilege definition

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7067/)

legrego committed:
improve 'create space privilege' button; add warning when creating global privilege

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7069/)

legrego committed:
update import reference

legrego committed:
Improve i18n

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7083/)

legrego committed:
a11y updates

legrego committed:
Adds unit tests for GET /api/security/privileges route

legrego committed:
adds functional API tests for GET /api/security/privileges

legrego committed:
implement open TODOs from design edits

legrego committed:
removes unnecessary isGroupLabelOption prop; fixes react warnings

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7130/)

legrego committed:
Support xpack.spaces.enabled: false

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7153/)

legrego committed:
additional tests

legrego committed:
remove unnecessary 'rbacEnabled' prop from EditRolePage

legrego committed:
update test: derive version action string

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7244/)

legrego committed:
disable space base privileges when not allowed

legrego committed:
hide 'change all' option when form is disabled

legrego committed:
improve space base privilege display when influenced by global base privileges

legrego committed:
Adds i18n support for the SimplePrivilegeSection

legrego committed:
remove debug code

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7262/)

legrego committed:
additional cleanup

legrego commented:
@cchaos I _think_ I solved almost everything from your [first design review](https://github.com/legrego/kibana/pull/13).

A couple of notes:
1) The button groups don't have disabled entries yet, so buttons will appear clickable even if they shouldn't be. Once the updated EUI is merged into Kibana, then I'll retest to make sure this works as expected.
2) I started doing accessibility testing, but I'm not able to get the privilege matrix or the flyout form to read the tables/button groups with enough context. The button groups are read, but I can't get the screen reader to announce which feature the button group applies to. For the matrix, I can't get the column headings to read, so the screen reader might announce "Discover Read All", but I have no idea which spaces "Read" or "All" is for.


One item I'm struggling with is the privilege matrix. When the matrix has a decent number of entries, the columns get squished, so the space avatars/links get cutoff:

![image](https://user-images.githubusercontent.com/3493255/51195683-976d8980-18bb-11e9-8e87-bf5fa6912447.png)

I don't think setting `maxWidth={false}` on the modal is the right way to go, and I'm hesitant to hardcode a pixel value here. Do you have any suggestions?

legrego committed:
Merge branch 'granular-app-privileges' into gap/security-ui

legrego committed:
snapshot updates

legrego committed:
improve type checks for components using initialFocus

legrego committed:
derive version for privileges functional api test

legrego committed:
Removes space privileges (and warns) when updating a role with spaces disabled

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7324/)

legrego committed:
fix visualize test

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7333/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7347/)

cchaos commented:
Another PR4U: https://github.com/legrego/kibana/pull/14

cchaos committed:
Better matrix layout (#14)

legrego commented:
Thanks @cchaos! Awesome stuff!

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7541/)

kobelb commented on x-pack/plugins/security/public/lib/effective_privileges/effective_privileges_utils.ts:
```diff
@@ -0,0 +1,64 @@
+import _ from 'lodash';
+
+// This is just a named set of actions.
+export interface ActionSet {
```
nit: I don't think we need this here as the only use is by `PrivilegeScenario` which extends from `ActionSet`, can we just remove this and add these members to `PrivilegeScenario`?

kobelb commented on x-pack/plugins/security/public/lib/privilege_utils.ts:
```diff
@@ -0,0 +1,27 @@
+ * Determines if the passed privilege spec defines global privileges.
+ * @param privilegeSpec
+ */
+export function isGlobalPrivilegeDefinition(privilegeSpec: KibanaPrivilegeSpec): boolean {
```
How would you feel about moving these to the models, so that we can more easily ask questions like this?

kobelb commented on x-pack/plugins/security/public/lib/role_utils.ts:
```diff
@@ -4,8 +4,8 @@
 
-import { get } from 'lodash';
-import { Role } from '../../common/model/role';
+import { cloneDeep, get } from 'lodash';
```
Same question about moving these to the "model"

legrego commented on x-pack/plugins/security/public/views/management/edit_role/components/privileges/kibana/feature_table/feature_table.tsx:
```diff
@@ -0,0 +1,286 @@
+        defaultMessage: 'Feature',
+      }),
+      render: (feature: TableFeature) => {
+        const tooltips = Object.entries(feature.privileges).reduce(
```
I'd like that very much ❤️

legrego commented on x-pack/plugins/security/public/views/management/edit_role/components/privileges/kibana/simple_privilege_section/simple_privilege_section.tsx:
```diff
@@ -0,0 +1,307 @@
+              onChange={this.onKibanaPrivilegeChange}
+              options={[
+                {
+                  value: NO_PRIVILEGE_VALUE,
```
Just an oversight, I'll clean this up

legrego commented on x-pack/plugins/security/public/views/management/edit_role/components/privileges/kibana/space_aware_privilege_section/privilege_space_table.tsx:
```diff
@@ -0,0 +1,306 @@
+            const explanation = effectivePrivileges.explainActualSpaceBasePrivilege(
+              record.spacesIndex
+            );
+            const canDisplayCustomize = hasCustomizations && explanation.privilege !== 'all';
```
We were attempting to be more technically correct to the user. If this entry had customizations, _but_ the effective privilege is `all`, then we know that the customizations have no impact whatsoever. So the most correct privilege to show is `all`.

I'll see if I can remove this and the other hard-coded checks in the effective privileges rewrite.

legrego commented on x-pack/plugins/security/public/views/management/edit_role/components/privileges/kibana/space_aware_privilege_section/space_aware_privilege_section.tsx:
```diff
@@ -0,0 +1,290 @@
+                  <strong>
+                    <FormattedMessage
+                      id="xpack.security.management.editRole.spaceAwarePrivilegeForm.kibanaUserTitle"
+                      defaultMessage="kibana_user"
```
You're probably right, but I'm tempted to leave this alone for now because it's not part of my PR, and was done by the `i18n` team. It shows up as an addition because I moved this entire file.

legrego commented on x-pack/plugins/security/public/views/management/edit_role/lib/transform_role_for_save.ts:
```diff
@@ -0,0 +1,34 @@
+import { IndexPrivilege, Role } from '../../../../../common/model';
+import { isGlobalPrivilegeDefinition } from '../../../../lib/privilege_utils';
+
+export function transformRoleForSave(role: Role, spacesEnabled: boolean) {
```
> However, transformRoleForSave maintains the Role interface while saveRole does not, was that the intent with the separation of these two?

I'd like to give myself that much credit, but that wasn't the case. We had the `role.elasticsearch.*` transforms within a React component before, and when I needed to add the `role.kibana.*` transforms, I decided to pull that logic out of the React component and consolidate it here. At the time, I didn't realize/forgot that we also had the other function which stripped out those other fields.

How do you feel about including the `omit` logic from https://github.com/elastic/kibana/blob/master/x-pack/plugins/security/public/objects/lib/roles.ts#L12 here, and having this function called from there instead of the react component? I'm ok updating this so that it no longer maintains the `Role` interface.

kobelb replied:
Your proposal sounds perfectly fine to me!

kobelb committed:
Apply suggestions from code review

Co-Authored-By: legrego <lgregorydev@gmail.com>

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7935/)

legrego committed:
Replace 'effective privileges' with 'kibana privilege calculator': WIP

legrego committed:
Replaces 'custom' hard-coded strings with a dedicated constant

legrego committed:
cleanup privilege display

legrego committed:
improve privilege display

legrego committed:
Merge branch 'gap/security-ui' of github.com:legrego/kibana into gap/security-ui

legrego commented on x-pack/plugins/security/server/routes/api/public/privileges/get.ts:
```diff
@@ -0,0 +1,26 @@
+
+import { PrivilegeMap } from 'x-pack/plugins/security/common/model';
+
+export function initGetPrivilegesApi(
```
My only reason to make this public is to enable API consumers to get a list of available privileges. Otherwise our role management API is going to be rather difficult to use without a bit of reverse-engineering on their side. I do agree that they don't need to be concerned with the "actions" though.

Do you think we should have a separate public API to enumerate just the privileges? The only downside there is that we wouldn't be using it at all, it would only be for other API consumers.

kobelb replied:
> Do you think we should have a separate public API to enumerate just the privileges? The only downside there is that we wouldn't be using it at all, it would only be for other API consumers.

The only other idea that I have is to use a flag to return the action information, which we could take advantage of internally, but not document for "public consumption".

legrego replied:
I'll do that!

legrego commented on x-pack/plugins/security/public/views/management/edit_role/components/privileges/kibana/space_aware_privilege_section/space_aware_privilege_section.tsx:
```diff
@@ -0,0 +1,290 @@
+    return [this.globalSpaceEntry, ...this.props.spaces];
+  };
+
+  private getAvailableSpaces = () => {
```
Thanks for the suggestion, this is _much_ cleaner! I'll update this

legrego committed:
remove unused imports/props/vars

legrego committed:
consolidate available spaces logic

legrego committed:
update tests to match description

legrego committed:
remove unused ActionSet interface

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8323/)

legrego committed:
Don't include privilege actions by default, but instead include them when requested via query parameter

kobelb committed:
Moving where the privileges tooltips are specified (#15)

legrego committed:
Merge branch 'granular-app-privileges' into gap/security-ui

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8342/)

legrego committed:
fix i18n lint issues

legrego committed:
more i18n lint fixes

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8351/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8355/)

legrego committed:
cleanup, bug fixes, and additional testing for privilege calculator

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8369/)

legrego commented:
@kobelb I've addressed most of your feedback, with the exception of:

1) Naming & Organization -- we know this needs work yet. We can either collaborate on this PR, or do a followup PR. I think my preference right now is a followup PR, so that we can do this work in a less chaotic place with usable diffs.

2) Privileges that don't follow the base -> feature hierarchy. We started discussing this at the end of last week, but this definitely needs some more thought. We could also address this in a followup if you want, but I don't have a strong opinion here.

3) This https://github.com/elastic/kibana/pull/26840#discussion_r249032257

kobelb commented:
> Naming & Organization -- we know this needs work yet. We can either collaborate on this PR, or do a followup PR. I think my preference right now is a followup PR, so that we can do this work in a less chaotic place with usable diffs.

Yeah, let's do this in a separate PR. 

> Privileges that don't follow the base -> feature hierarchy. We started discussing this at the end of last week, but this definitely needs some more thought. We could also address this in a followup if you want, but I don't have a strong opinion here.

Let's briefly chat about this and write some notes down (while it's still somewhat fresh in our brains), so we're prepared for this possible future, but I don't think it should block this PR.

I'll give everything else a quick look over shortly.

legrego committed:
consolidate transform role logic

legrego committed:
remove unused import

legrego committed:
Merge branch 'granular-app-privileges' into gap/security-ui

legrego committed:
test fix: update expected privilege action mapping

legrego committed:
i18n linter fixes

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9120/)

legrego committed:
update security scenarios to reflect changed feature privileges

legrego commented on x-pack/plugins/security/public/lib/kibana_privilege_calculator/kibana_privilege_calculator_types.ts:
```diff
@@ -0,0 +1,58 @@
+  actualPrivilege: string;
+  actualPrivilegeSource: PRIVILEGE_SOURCE;
+  isDirectlyAssigned: boolean;
+  supercededPrivilege?: string;
```
You have no idea how many times I've done that. I keep wanting to spell it with a **c**

legrego committed:
simplify empty role via buildRole()

legrego committed:
fix spelling: superceded => superseded

legrego committed:
I speel gud

legrego committed:
fix global find/replace error

legrego committed:
adds tests against the public calculateAllowedPrivileges function

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9349/)

legrego commented:
retest

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9361/)

kobelb commented:
If I fill out the index (don't press return) hit `TAB` twice, fill out the privileges hit `TAB` twice, it'll crash with the following:

![tab-crash](https://user-images.githubusercontent.com/627123/51868326-9c422c80-2302-11e9-83b9-afb234707014.gif)

kobelb commented on x-pack/plugins/security/public/views/management/edit_role/components/privileges/kibana/simple_privilege_section/unsupported_space_privileges_warning.tsx:
```diff
@@ -0,0 +1,24 @@
+    return (
+      <FormattedMessage
+        id="xpack.security.management.editRole.simplePrivilegeForm.unsupportedSpacePrivilegesWarning"
+        defaultMessage="This role contains privilege definitons for spaces, but spaces are not enabled in Kibana. Saving this role will remove these privileges."
```
nit: 
```suggestion
        defaultMessage="This role contains privilege definitions for spaces, but spaces are not enabled in Kibana. Saving this role will remove these privileges."
```

kobelb commented:
When creating a role which only grants Elasticsearch index privileges with spaces disabled, when you go to save a validation error is preventing it from saving:

<img width="929" alt="screen shot 2019-01-28 at 2 04 52 pm" src="https://user-images.githubusercontent.com/627123/51871298-d9122180-230a-11e9-8219-5b9bccfac69f.png">

kobelb committed:
Update x-pack/plugins/security/public/views/management/edit_role/components/privileges/kibana/simple_privilege_section/unsupported_space_privileges_warning.tsx

Co-Authored-By: legrego <lgregorydev@gmail.com>

legrego committed:
allow roles to be created without kibana privileges

legrego committed:
Merge branch 'granular-app-privileges' into gap/security-ui

legrego committed:
Merge branch 'gap/security-ui' of github.com:legrego/kibana into gap/security-ui

kobelb committed:
Specifying a privileges tooltip for APM (#16)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9403/)

legrego committed:
update failing snapshots

legrego committed:
Add expected privilege/action mappings for infrastructure-ui-source saved object types

legrego committed:
Merge branch 'gap/security-ui' of github.com:legrego/kibana into gap/security-ui

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9476/)
